admin/server/device_code.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// handleDeviceCodeRequest creates a 24 digit random device code and 8 digit user code and returns that
+// to the client. The device code is used to poll for an access token, and the user code is displayed
+// to the user and is used to authorize the device. The device code and user code are stored in the
+// server's device code store.
+func (s *Server) handleDeviceCodeRequest(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to read request body: %w", err))
+		return
+	}
+	bodyStr := string(body)
+	values, err := url.ParseQuery(bodyStr)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	clientID := values.Get("client_id")
+	if clientID == "" {
+		http.Error(w, "client_id is required", http.StatusBadRequest)
+		return
+	}
+	scopes := strings.Split(values.Get("scope"), " ")
+	if len(scopes) == 0 {
+		http.Error(w, "scope is required", http.StatusBadRequest)
+		return
+	}
+	if len(scopes) > 1 || scopes[0] != "full_account" {
+		http.Error(w, "invalid scope", http.StatusBadRequest)
+		return
+	}
+	authCode, err := s.admin.IssueAuthCode(r.Context(), clientID)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to issue auth code: %w", err))
+		return
+	}
+
+	verificationURI := s.opts.DeviceVerificationHost + "/oauth/device"
+	resp := DeviceCodeResponse{
+		DeviceCode:              authCode.DeviceCode,
+		UserCode:                authCode.UserCode,
+		VerificationURI:         verificationURI,
+		VerificationCompleteURI: verificationURI + "?user_code=" + authCode.UserCode,
+		ExpiresIn:               tokenExpirationSeconds,
+		PollingInterval:         5,
+	}
+	e := json.NewEncoder(w)
+	e.SetIndent("", "  ")
+	err = e.Encode(resp)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to encode response: %w", err))
+	}
+}
+
+// handleUserCodeConfirmation handles the user code confirmation page. The user code is displayed
+// to the user and they are asked to confirm that they want to authorize the device. If the user
+// confirms, the device code is marked as approved in the server's device code store.
+func (s *Server) handleUserCodeConfirmation(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	userCode := r.URL.Query().Get("user_code")
+	if userCode == "" {
+		http.Error(w, "user_code is required", http.StatusBadRequest)
+		return
+	}
+	confirmation := r.URL.Query().Get("code_confirmed")
+	if confirmation == "" {
+		http.Error(w, "no code confirmation", http.StatusBadRequest)
+		return
+	}
+
+	claims := auth.GetClaims(r.Context())
+	if claims == nil {
+		internalServerError(w, fmt.Errorf("did not find any claims, %w", errors.New("server error")))
+		return
+	}
+	userID := claims.OwnerID()
+	if userID == "" {
+		internalServerError(w, fmt.Errorf("did not find user id in claims, %w", errors.New("server error")))
+		return
+	}
```

Comment:
Should check `claims.OwnerType() == auth.OwnerTypeUser` instead. Then `OwnerID()` is guaranteed to not be empty